### ISSUE TITLE
Not support subquery？

### ISSUE THREAD

**Issue by wonb168** [reata/sqllineage#192]
with c1 as (select * from ttt),
c2 as(select * from c1)
select * from (select * from c2 as aa)b

I want the table lineage inclued subquery name, this example need output:
ttt -> c1
c1 -> c2
c2 -> b

**Comment by reata** [comment]
I see why you would want this feature, but unfortunately this would need a lot of changes to current code base. To make it happen, we need to refactor the code not limited to the following aspect:

- Fix column lineage tracing for CTE referencing each other. Here CTE c2 is using data from CTE c1. But our lineage says only "tab1.* <- b.* <- <default>.c2.*", rather than "tab1.* <- b.* <- <default>.c2.* <- <default>.c1.*". That's definitely a bug.
- Using the same way to build table level lineage as we do for column. Right now, each SQL statement is exposing table lineage info like read, write, cte, etc. Instead, we need a graph. (Internally it is already stored in a graph, but we're not leveraging it).
- Allow table lineage to be presented without a target table. Because current design only draws the DAG if there's a target table. I'm actually not convinced that we should change this.

In conclusion, thanks for reporting this, but this may take time.

**Comment by wonb168** [comment]
Because complex buniess login need many steps：temp table/subquery/with. 
And then draw a diagraph to view the whole relations, the graph only have source table and target in unenoughh.
Another example:
create temporary table tmp_replenish_transfer_level as
with template_info as (
select id::text as biz_action_template_code
,parameter_table_id
from tenant_${tenant_id}_biz.gto_biz_action_template
), parameter_table_info as (
select a.id,b.biz_action_template_code
from tenant_${tenant_id}_biz.base_parameter_table as a
inner join template_info as b
on a.id = b.parameter_table_id
where a.is_deleted = 0
)
select biz_action_template_code
,min(id) as id,min(parameter_value_id) as parameter_value_id,min(parameter_code) as parameter_code,min(parameter_value::text) as parameter_value
,coalesce(min(replenish_transfer_level::text),'operating_unit') as replenish_transfer_level
from (
select a.biz_action_template_code
,a.id, b.id as parameter_value_id, b.parameter_code, b.parameter_value
, b.parameter_value ->> 'gto.replenish_transfer_level' as replenish_transfer_level
from parameter_table_info as a
inner join tenant_${tenant_id}_biz.base_parameter_value as b
on a.id = b.parameter_table_id
where b.parameter_code = 'gto.replenish_transfer_level' and b.is_deleted = 0
union all
select distinct id::varchar as biz_action_template_code
,null::int as id,null::int as parameter_value_id,null as parameter_code,null::jsonb as parameter_value
,null as replenish_transfer_level
from tenant_${tenant_id}_biz.gto_biz_action_template
where is_deleted='0' and is_enable='1'
) aa
group by biz_action_template_code
order by biz_action_template_code::int

**Comment by reata** [comment]
Thanks for the example. I'll prioritize this issue. 
Maybe I ask what's the tech stack / sql dialect you're using? Looks like PostgreSQL.

**Comment by reata** [comment]
The first step is done after we close #196

**Comment by wonb168** [comment]
what's the tech stack / sql dialect you're using? Looks like PostgreSQL.
right, it's postgresql

**Comment by wonb168** [comment]
do you have a wechat or other social account？maybe communicate needs more convenient

**Comment by reata** [comment]
> do you have a wechat or other social account？maybe communicate needs more convenient

I'd prefer to keep all the discussion here so that later people know how decision is made along the way. But feel free to email me if that's convenient for you.

**Comment by wonb168** [comment]
how about this issue going?

**Comment by reata** [comment]
> how about this issue going?

Not so fast. Right now I'm still focusing on fixing various edge case bugs for column level lineage. Might need another 2 or 3 minor version release before things are stable at column level.

**Comment by wonb168** [comment]
column lineage can show subquery , then table lineage contain subquery perhaps easier, for example directly from column lineage is one way.

**Comment by huanglaoxie0503** [comment]
how about this issue going?

**Comment by reata** [comment]
We're prioritizing #302, changing the underlying parser to support more SQL dialects. Every other issue is put on hold for the moment.

**Comment by BrandenXia** [comment]
how about this issue going?

**Comment by reata** [comment]
> how about this issue going?

Can you share the use case? I really didn't anticipate this feature should attract so many attentions?

**Comment by BrandenXia** [comment]
> > how about this issue going?
> 
> Can you share the use case? I really didn't anticipate this feature should attract so many attentions?

I'm now using MySQL. When doing data analysis, we use a lot of subquery. When a problem occur, I use this library to analyze the SQL so that we trace the origin of each field. However, when meeting subqueries, it's not so useful because we can only see that the data starts from a subquery.
Now we use sqlparse to do some preprocessing to the SQL to expand the subquery before we put it into LineageRunner.

**Comment by reata** [comment]
@BrandenXia it feels to me that your problem is mostly related with incomplete column lineage. Can you check #303 to see if that's the same with your issue? Or you can share some queries that you don't get what you expect with column lineage.

**Comment by BrandenXia** [comment]
> @BrandenXia it feels to me that your problem is mostly related with incomplete column lineage. Can you check #303 to see if that's the same with your issue? Or you can share some queries that you don't get what you expect with column lineage.

Yeah, that's basically same with the problem I mentioned.
